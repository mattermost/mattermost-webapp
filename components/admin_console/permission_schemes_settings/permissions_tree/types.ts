// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type Permission = {
    id: string;
    combined?: boolean;
    permissions: string[];
}
export type Group = {
    id: string;
    permissions: Array<Permission | string>;
}

export type AdditionalValues = {
    edit_post: {
        editTimeLimitButton: JSX.Element;
    };
}
