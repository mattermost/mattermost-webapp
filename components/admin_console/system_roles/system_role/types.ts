// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type ReadAccess = 'read';
export const readAccess: ReadAccess = 'read';
export type WriteAccess = 'write';
export const writeAccess: WriteAccess = 'write';
export type NoAccess = false;
export const noAccess: NoAccess = false;
export type MixedAccess = 'mixed';
export const mixedAccess: MixedAccess = 'mixed';

export type PermissionAccess = ReadAccess | WriteAccess | NoAccess;
export type PermissionsToUpdate = Record<string, ReadAccess | WriteAccess | NoAccess>;
export type PermissionToUpdate = {
    name: string;
    value: PermissionAccess;
};

export type SystemSection = {
    name: string;
    hasDescription?: boolean;
    subsections?: SystemSection[];
    disabled?: boolean;
}
