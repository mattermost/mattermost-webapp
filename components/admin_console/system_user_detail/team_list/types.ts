// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Team, TeamMembership} from '@mattermost/types/teams';

export type TeamWithMembership = Team & TeamMembership;
