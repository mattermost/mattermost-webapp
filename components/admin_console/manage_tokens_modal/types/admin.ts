// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// this patches the wrong type definition in node_modules/mattermost-redux/src/types/admin.ts

import {AdminState} from 'mattermost-redux/src/types/admin';
import {Dictionary, RelationOneToOne} from 'mattermost-redux/src/types/utilities';
import {UserProfile} from 'mattermost-redux/types/users';
import {UserAccessToken} from 'mattermost-redux/src/types/users';

export type AdminTokenState = AdminState & {

    // userAccessTokensForUser in definition rather than userAccessTokensByUser
    userAccessTokensByUser?: RelationOneToOne<UserProfile, Dictionary<UserAccessToken>>;
};
