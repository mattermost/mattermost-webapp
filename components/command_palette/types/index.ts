// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum CommandPaletteEntities {
    Channel = 'channel',
    Playbooks = 'playbooks',
    Boards = 'boards',
    Messages = 'messages',
    Files = 'files',
    Runs = 'runs',
    GoTo = 'goTo',
}

export enum ChannelType {
    DM_CHANNEL= 'D',
    GM_CHANNEL= 'G',
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum BoardsType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum PlaybooksType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

