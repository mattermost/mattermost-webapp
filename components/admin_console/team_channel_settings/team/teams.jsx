// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {createSelector} from 'reselect';

import {getTeams as fetchTeams} from 'mattermost-redux/actions/teams';
import {getTeams} from 'mattermost-redux/selectors/entities/teams';

import {t} from 'utils/i18n';

import TeamList from './team_list.jsx';

const getSortedListOfTeams = createSelector(
    getTeams,
    (teams) => Object.values(teams).sort((a, b) => a.name.localeCompare(b.name))
);

function mapStateToProps(state) {
    return {
        data: getSortedListOfTeams(state),
        total: 100,
        emptyListTextId: t('admin.team_settings.team_list.no_teams_found'),
        emptyListTextDefaultMessage: 'No teams found',
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            getData: fetchTeams,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(TeamList);
