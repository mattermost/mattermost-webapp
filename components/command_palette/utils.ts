// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Channel} from '@mattermost/types/channels';
import {Team} from '@mattermost/types/teams';

// See LICENSE.txt for license information.
import {CommandPaletteItem} from './command_palette_list_item/command_palette_list_item';
import {
    CmdPalettePictographIcon,
    CmdPalettePictographType,
} from './command_palette_list_item/command_palette_list_item_pictograph';
import {ChannelType, CommandPaletteEntities} from './types';

const channelTypeOrder = {
    [ChannelType.OPEN_CHANNEL]: 0,
    [ChannelType.PRIVATE_CHANNEL]: 1,
    [ChannelType.DM_CHANNEL]: 2,
    [ChannelType.GM_CHANNEL]: 3,
};

function filterName(name: string): string {
    return name.replace(/[.,'"\/#!$%\^&\*;:{}=\-_`~()]/g, ''); // eslint-disable-line no-useless-escape
}

export function channelToCommandPaletteItemTransformer(channels: Channel[], teams: Record<string, Team>): CommandPaletteItem[] {
    return channels.map((channel) => {
        const isArchived = channel.delete_at ? channel.delete_at !== 0 : false;

        let channelIcon = CmdPalettePictographIcon.GLOBE;
        if (isArchived) {
            channelIcon = CmdPalettePictographIcon.ARCHIVE;
        }
        return {
            description: channel.type === ChannelType.DM_CHANNEL ? channel.name : '',
            id: channel.id,
            isDeactivated: isArchived,
            teamName: teams[channel.team_id]?.name,
            pictograph: {
                type: CmdPalettePictographType.ICON,
                pictographItem: channelIcon,
            },
            subType: channel.type as ChannelType,
            title: channel.display_name,
            type: CommandPaletteEntities.Channel,
            isArchived,
        };
    });
}

function sortChannelsByTypeAndDisplayName(locale: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem): number {
    if (channelTypeOrder[cmdPaletteItemA.subType] !== channelTypeOrder[cmdPaletteItemB.subType]) {
        if (channelTypeOrder[cmdPaletteItemA.subType] < channelTypeOrder[cmdPaletteItemB.subType]) {
            return -1;
        }

        return 1;
    }

    const aDisplayName = filterName(cmdPaletteItemA.title);
    const bDisplayName = filterName(cmdPaletteItemB.title);

    if (aDisplayName !== bDisplayName) {
        return aDisplayName.toLowerCase().localeCompare(bDisplayName.toLowerCase(), locale, {numeric: true});
    }

    return cmdPaletteItemA.title.toLowerCase().localeCompare(cmdPaletteItemA.title.toLowerCase(), locale, {numeric: true});
}

function sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.lastViewedAt && cmdPaletteItemB.lastViewedAt) {
        return cmdPaletteItemB.lastViewedAt - cmdPaletteItemA.lastViewedAt;
    } else if (cmdPaletteItemA.lastViewedAt) {
        return -1;
    } else if (cmdPaletteItemB.lastViewedAt) {
        return 1;
    }

    // MM-12677 When this is migrated this needs to be fixed to pull the user's locale
    return sortChannelsByTypeAndDisplayName('en', cmdPaletteItemA, cmdPaletteItemB);
}

export function commandPaletteSorter(prefix: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.isArchived && !cmdPaletteItemB.isArchived) {
        return 1;
    } else if (!cmdPaletteItemA.isArchived && cmdPaletteItemB.isArchived) {
        return -1;
    }

    if (cmdPaletteItemA.isDeactivated && !cmdPaletteItemB.isDeactivated) {
        return 1;
    } else if (cmdPaletteItemB.isDeactivated && !cmdPaletteItemA.isDeactivated) {
        return -1;
    }

    let aDisplayName = cmdPaletteItemA.title.toLowerCase();
    let bDisplayName = cmdPaletteItemA.title.toLowerCase();

    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.DM_CHANNEL &&
        aDisplayName.startsWith('@')) {
        aDisplayName = aDisplayName.substring(1);
    }

    if (cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.DM_CHANNEL && bDisplayName.startsWith('@')) {
        bDisplayName = bDisplayName.substring(1);
    }

    const aStartsWith = aDisplayName.startsWith(prefix) || cmdPaletteItemA.description.toLowerCase().startsWith(prefix);
    const bStartsWith = bDisplayName.startsWith(prefix) || cmdPaletteItemB.description.toLowerCase().startsWith(prefix);

    // Open channels user haven't interacted should be at the  bottom of the list
    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.OPEN_CHANNEL &&
        !cmdPaletteItemA.lastViewedAt &&
        ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
            cmdPaletteItemB.subType !== ChannelType.OPEN_CHANNEL) || cmdPaletteItemB.lastViewedAt)) {
        return 1;
    } else if ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.OPEN_CHANNEL) || !cmdPaletteItemB.lastViewedAt) {
        return -1;
    }

    // Sort channels starting with the search term first
    if (aStartsWith && !bStartsWith) {
        return -1;
    } else if (!aStartsWith && bStartsWith) {
        return 1;
    }
    return sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA, cmdPaletteItemB);
}
