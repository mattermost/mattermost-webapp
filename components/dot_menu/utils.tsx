// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {trackEvent} from 'actions/telemetry_actions';

import {EventTypes, TELEMETRY_CATEGORIES} from 'utils/constants';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (e: ChangeEvent, suffix: string): void => {
    if (e.type === EventTypes.CLICK) {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.CLICK + '_' + suffix);
    } else {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.SHORTCUT + '_ ' + suffix);
    }
};
