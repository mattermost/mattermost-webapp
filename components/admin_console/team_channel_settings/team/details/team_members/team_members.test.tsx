// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {shallow} from 'enzyme';

import {Team, TeamMembership} from 'mattermost-redux/types/teams';
import {UserProfile} from 'mattermost-redux/types/users';

import {TestHelper} from '../../../../../../utils/test_helper';

import TeamMembers from './team_members';

describe('admin_console/team_channel_settings/team/TeamMembers', () => {
    const user1: UserProfile = Object.assign(TestHelper.getUserMock({id: 'user-1'}));
    const membership1: TeamMembership = Object.assign(TestHelper.getTeamMembershipMock({user_id: 'user-1'}));
    const user2: UserProfile = Object.assign(TestHelper.getUserMock({id: 'user-2'}));
    const membership2: TeamMembership = Object.assign(TestHelper.getTeamMembershipMock({user_id: 'user-2'}));
    const user3: UserProfile = Object.assign(TestHelper.getUserMock({id: 'user-3'}));
    const membership3: TeamMembership = Object.assign(TestHelper.getTeamMembershipMock({user_id: 'user-3'}));
    const team: Team = Object.assign(TestHelper.getTeamMock({id: 'team-1'}));

    const baseProps = {
        teamId: 'team-1',
        team,
        users: [user1, user2, user3],
        usersToRemove: {},
        usersToAdd: {},
        teamMembers: {
            [user1.id]: membership1,
            [user2.id]: membership2,
            [user3.id]: membership3,
        },

        totalCount: 3,
        loading: false,
        searchTerm: '',
        onAddCallback: jest.fn(),
        onRemoveCallback: jest.fn(),
        updateRole: jest.fn(),

        actions: {
            getTeamStats: jest.fn(),
            loadProfilesAndReloadTeamMembers: jest.fn(),
            searchProfilesAndTeamMembers: jest.fn(),
            setModalSearchTerm: jest.fn(),
        },
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <TeamMembers {...baseProps}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot loading no users', () => {
        const wrapper = shallow(
            <TeamMembers
                {...baseProps}
                users={[]}
                teamMembers={{}}
                totalCount={0}
                loading={true}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });
});