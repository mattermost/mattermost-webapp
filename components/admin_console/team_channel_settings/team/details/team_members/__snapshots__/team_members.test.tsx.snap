// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin_console/team_channel_settings/team/TeamMembers should match snapshot 1`] = `
<AdminPanel
  button={
    <ModalToggleButton
      className="btn btn-primary"
      dialogProps={
        Object {
          "excludeUsers": Object {},
          "filterExcludeGuests": true,
          "includeUsers": Object {},
          "onAddCallback": [Function],
          "skipCommit": true,
          "team": Object {
            "allow_open_invite": false,
            "allowed_domains": "",
            "company_name": "",
            "create_at": 0,
            "delete_at": 0,
            "description": "",
            "display_name": "name",
            "email": "",
            "group_constrained": false,
            "id": "team-1",
            "invite_id": "",
            "name": "DN",
            "scheme_id": "id",
            "type": "O",
            "update_at": 0,
          },
        }
      }
      dialogType={[Function]}
      id="addTeamMembers"
    >
      <FormattedMessage
        defaultMessage="Add Members"
        id="admin.team_settings.team_details.add_members"
        values={Object {}}
      />
    </ModalToggleButton>
  }
  className=""
  id="teamMembers"
  subtitleDefault="A list of users who are currently in the team right now"
  subtitleId="admin.team_settings.team_detail.membersDescription"
  titleDefault="Members"
  titleId="admin.team_settings.team_detail.membersTitle"
>
  <UserGrid
    excludeUsers={Object {}}
    includeUsers={Object {}}
    loadPage={[Function]}
    loading={true}
    memberships={
      Object {
        "user-1": Object {
          "delete_at": 0,
          "mention_count": 0,
          "msg_count": 0,
          "roles": "team_user",
          "scheme_admin": false,
          "scheme_user": true,
          "team_id": "team_id",
          "user_id": "user-1",
        },
        "user-2": Object {
          "delete_at": 0,
          "mention_count": 0,
          "msg_count": 0,
          "roles": "team_user",
          "scheme_admin": false,
          "scheme_user": true,
          "team_id": "team_id",
          "user_id": "user-2",
        },
        "user-3": Object {
          "delete_at": 0,
          "mention_count": 0,
          "msg_count": 0,
          "roles": "team_user",
          "scheme_admin": false,
          "scheme_user": true,
          "team_id": "team_id",
          "user_id": "user-3",
        },
      }
    }
    removeUser={[Function]}
    scope="team"
    search={[Function]}
    term=""
    totalCount={3}
    updateMembership={[Function]}
    users={
      Array [
        Object {
          "auth_data": "",
          "auth_service": "",
          "create_at": 0,
          "delete_at": 0,
          "email": "",
          "email_verified": true,
          "first_name": "",
          "id": "user-1",
          "is_bot": false,
          "last_name": "",
          "last_picture_update": 0,
          "locale": "",
          "nickname": "",
          "notify_props": Object {
            "channel": "false",
            "comments": "never",
            "desktop": "default",
            "desktop_sound": "false",
            "email": "false",
            "first_name": "false",
            "mark_unread": "mention",
            "mention_keys": "",
            "push": "none",
            "push_status": "offline",
          },
          "position": "",
          "roles": "",
          "terms_of_service_create_at": 0,
          "terms_of_service_id": "",
          "update_at": 0,
          "username": "some-user",
        },
        Object {
          "auth_data": "",
          "auth_service": "",
          "create_at": 0,
          "delete_at": 0,
          "email": "",
          "email_verified": true,
          "first_name": "",
          "id": "user-2",
          "is_bot": false,
          "last_name": "",
          "last_picture_update": 0,
          "locale": "",
          "nickname": "",
          "notify_props": Object {
            "channel": "false",
            "comments": "never",
            "desktop": "default",
            "desktop_sound": "false",
            "email": "false",
            "first_name": "false",
            "mark_unread": "mention",
            "mention_keys": "",
            "push": "none",
            "push_status": "offline",
          },
          "position": "",
          "roles": "",
          "terms_of_service_create_at": 0,
          "terms_of_service_id": "",
          "update_at": 0,
          "username": "some-user",
        },
        Object {
          "auth_data": "",
          "auth_service": "",
          "create_at": 0,
          "delete_at": 0,
          "email": "",
          "email_verified": true,
          "first_name": "",
          "id": "user-3",
          "is_bot": false,
          "last_name": "",
          "last_picture_update": 0,
          "locale": "",
          "nickname": "",
          "notify_props": Object {
            "channel": "false",
            "comments": "never",
            "desktop": "default",
            "desktop_sound": "false",
            "email": "false",
            "first_name": "false",
            "mark_unread": "mention",
            "mention_keys": "",
            "push": "none",
            "push_status": "offline",
          },
          "position": "",
          "roles": "",
          "terms_of_service_create_at": 0,
          "terms_of_service_id": "",
          "update_at": 0,
          "username": "some-user",
        },
      ]
    }
  />
</AdminPanel>
`;

exports[`admin_console/team_channel_settings/team/TeamMembers should match snapshot loading no users 1`] = `
<AdminPanel
  button={
    <ModalToggleButton
      className="btn btn-primary"
      dialogProps={
        Object {
          "excludeUsers": Object {},
          "filterExcludeGuests": true,
          "includeUsers": Object {},
          "onAddCallback": [Function],
          "skipCommit": true,
          "team": Object {
            "allow_open_invite": false,
            "allowed_domains": "",
            "company_name": "",
            "create_at": 0,
            "delete_at": 0,
            "description": "",
            "display_name": "name",
            "email": "",
            "group_constrained": false,
            "id": "team-1",
            "invite_id": "",
            "name": "DN",
            "scheme_id": "id",
            "type": "O",
            "update_at": 0,
          },
        }
      }
      dialogType={[Function]}
      id="addTeamMembers"
    >
      <FormattedMessage
        defaultMessage="Add Members"
        id="admin.team_settings.team_details.add_members"
        values={Object {}}
      />
    </ModalToggleButton>
  }
  className=""
  id="teamMembers"
  subtitleDefault="A list of users who are currently in the team right now"
  subtitleId="admin.team_settings.team_detail.membersDescription"
  titleDefault="Members"
  titleId="admin.team_settings.team_detail.membersTitle"
>
  <UserGrid
    excludeUsers={Object {}}
    includeUsers={Object {}}
    loadPage={[Function]}
    loading={true}
    memberships={Object {}}
    removeUser={[Function]}
    scope="team"
    search={[Function]}
    term=""
    totalCount={0}
    updateMembership={[Function]}
    users={Array []}
  />
</AdminPanel>
`;
